### PR TITLE
Implement __SCOPE__

### LINKED ISSUE #775: [Feature request] A predeclared EQUS symbol for "the current global label"
For instance, `_GLOBAL` or `_SCOPE`. This would be undefined if there is no current globally scoped label.

This would be useful to avoid repeating names. I've had multiple use cases for code like this:

`Foobar:: mac Foobar`

In this case `mac` might be defining more labels that use `Foobar` as a prefix or suffix. I like to avoid having `mac` do `\1::` to declare the main label itself, since that makes `Foobar`'s definition harder to find.

Or, `mac` could be mentioning `Foobar` in some error message. This is useful for macros that don't declare any code/data/labels, but just check assertions about what's already there, like data table lengths or verifying that certain related symbols are defined.

### PR DESCRIPTION
**Pull request by Rangi42**
`__SCOPE__` expands to the currently scoped label. It is empty if inside the main scope.

Fixes #775

### CODE REVIEW

File: src/asm/symbol.c [hunk 1]
```diff
@@ -128,6 +128,22 @@ static char const *Callback__FILE__(void)
...
+static char const *Callback__SCOPE__(void)
+{
+	static char buf[MAXSYMLEN + 1];
+
+	if (!labelScope)
+		fatalerror("__SCOPE__ does not make sense in main scope\n");
+
+	size_t n = strlen(labelScope);
+
+	if (n > sizeof(buf) - 1)
```
Comment (Rangi42) [inline src/asm/symbol.c:140]: Can we assume `labelScope` will be a valid label, and omit this check?

**Comment by Rangi42** [comment]
`checkpatch` is complaining "WARNING: Move `const` after `static` - use '`static const char`'" regarding `static char const *Callback__SCOPE__(void)`, but I copied it from `static char const *Callback__FILE__(void)` which doesn't get this warning. :/

**Comment by ISSOtm** [comment]
As I said, I don't like this functionality because I think it caters to too specific a purpose, which natively handling in RGBASM brings little benefit. (Paging @AntonioND to have a third opinion)

Plus, erroring out when no scope is present means there's no way to handle that case, it should rather not be defined then, as the documentation suggests ("undefined").

**Comment by Rangi42** [comment]
When no scope is present, `__SCOPE__` *is* undefined: `DEF(__SCOPE__)` returns false, and evaluating `__SCOPE__` acts like evaluating any other undefined symbol (i.e. the error message depends on whether it's evaluated as a constant, or interpolated into a string). (Yes, it still has an entry in the symbol hashtable, but the lookup function explicitly checks for it. Compare with `_NARG`, which counts as defined even outside a macro, but causes an error when evaluated.)

File: src/asm/symbol.c (hunk 1 above)
Comment (ISSOtm) [reply src/asm/symbol.c:140]: It should be; I would still make this an `assert` because debug sanity checks don't hurt.

**Comment by ISSOtm** [comment]
> When no scope is present, `__SCOPE__` _is_ undefined: `DEF(__SCOPE__)` returns false, and evaluating `__SCOPE__` acts like evaluating any other undefined symbol (i.e. the error message depends on whether it's evaluated as a constant, or interpolated into a string).

Then what is [this check](https://github.com/gbdev/rgbds/pull/776/files#diff-4cf02ed4bfe24135ede79fefe2f3d1ece6d0edbec3b59de33aacbed8ec45b2ecR135) for? (I didn't look further after seeing it, assuming it was the only one; hence my inaccurate reply)

> (Yes, it still has an entry in the symbol hashtable, but the lookup function explicitly checks for it. Compare with `_NARG`, which counts as defined even outside a macro, but causes an error when evaluated.)

I don't like that it's hardcoded, tbh. `_NARG` is imo different because there's no reason to try using it outside of a macro; however, a macro may be written in a way to act differently whether it's in a scope or not.

**Comment by Rangi42** [comment]
The checks `if (!labelScope)` and `if (n > sizeof(buf) - 1)` are both checking for things that should never be the case, because they are avoided before calling `Callback__SCOPE__`. I'll make them assertions.

I agree that hardcoding the `!strcmp(name, "__SCOPE__")` check is not ideal, but removing and re-adding `__SCOPE__` to the hashtable would have the memory issues that you brought up.

**Comment by ISSOtm** [comment]
No memory issues at all, since the buffer returned is global. Removing and re-inserting the symbol manually into the hash table would be more problematic (repeated allocations and de-allocations), and it would still be hardcoding, which I really really want to avoid. Call it PTSD perhaps.

**Comment by Rangi42** [comment]
The repeated allocations and de-allocations are what I meant by memory issues.

To technically avoid hardcoding a `strcmp`, all `Symbol`s could gain a new property: either `bool isDefinedInMainScope`, or `bool (*defCallback)(void)`. (`sym_FindExactSymbol` would then do `return (!sym->defCallback || sym->defCallback()) ? sym : NULL;`.) The former feels like a cheat that's still hardcoding. The latter is a more flexible mechanism, but would still be created just for this one symbol. (Maybe it would have other uses?)

**Comment by ISSOtm** [comment]
The implications of the latter on the rest of the codebase's assumptions are not something I would want to deal with, certainly not in 0.5.0.

**Comment by Rangi42** [comment]
Actually, another solution would be to act like `PCSymbol`: store `__SCOPE__Symbol` as a static global, and then do `return (sym != __SCOPE__Symbol || labelScope) ? sym : NULL`. (Like how `sym_IsPC` just "hardcodes" `return sym == PCSymbol`.)

**Comment by ISSOtm** [comment]
The difference, again imo, is that PC is a necessary evil, but that this functionality isn't worth all that trouble.

**Comment by Rangi42** [comment]
Okay. For now I've updated this to use a `sym == __SCOPE__Symbol` as the simplest of those ways to arbitrarily check for a particular symbol.

Any idea why the `static const char` warnings are inconsistent?

**Comment by ISSOtm** [comment]
Not the slightest.

**Comment by aaaaaa123456789** [comment]
> As I said, I don't like this functionality because I think it caters to too specific a purpose, which natively handling in RGBASM brings little benefit.

Not really. This can be used by all sorts of metaprograming, ranging from good error messages to structure-defining macros.

See: `__FUNC__` in C (I think added in C99).

**Comment by ISSOtm** [comment]
Structure-defining macros exist of all sorts, and never needed this (nor has it been bothering); and what error messages are bad?

**Comment by aaaaaa123456789** [comment]
> what error messages are bad?

```
foo: MACRO
  ___foo \1, __SCOPE__
ENDM

; I'm sure you could do this with string
; interpolation, but I don't remember how to
___foo: MACRO
  if (\1) > 40
    warn "Foo value for \2 over threshold!
  endc
  db \1
ENDM
```

**Comment by ISSOtm** [comment]
There's never been a need for this, and it's incompatible with a variety of uses, which is why I'm deeming it as "not enough".

**Comment by Rangi42** [comment]
That would be:

```
foo: MACRO
  if (\1) > 40
    warn "Foo value for {__SCOPE__} over threshold!"
  endc
  db \1
ENDM

TrainerA:
    db "A-chan"
    foo 38
    bar 200

TrainerB:
    db "B-kun"
    foo 42 ; oops
    bar 180
```

**Comment by ISSOtm** [comment]
This would only be beneficial if there were more than a screenful of lines between the last global label (assuming none of the macros in the middle define one) and the erroring macro. Frankly, it's more fragile than anything.

**Comment by aaaaaa123456789** [comment]
You're assuming that people know how to read error messages. A week in pret's #pokecrystal will quickly prove you wrong.

**Comment by ISSOtm** [comment]
Nothing can be done for people who don't want to read error messages.

**Comment by Rangi42** [comment]
I'll go ahead and close this, but am still interested what @AntonioND thinks of the feature and of this implementation approach.

**Comment by AntonioND** [comment]
So I'm a bit confused, what's the purpose of this apart of having custom error messages?

**Comment by Rangi42** [comment]
It could also be used to make new labels based on the current one.

```
vec_struct: MACRO
{__SCOPE__}_X:: db
{__SCOPE__}_Y:: db
{__SCOPE__}_Z:: db
ENDM

wPosition:: vec_struct
```

This avoids having to either repeat a label (`wPosition:: vec_struct wPosition`, using `\1` instead of `__SCOPE__`) or make `vec_struct wFoo` define the top-level label `wFoo` as well.

**Comment by AntonioND** [comment]
Well, fair enough. In my case, I think that the amount of code to support this isn't that big, but at this point I really don't know the implications of supporting it (if it interacts with other features somehow). I guess we could just let users comment here if they want the feature?

**Comment by daid** [comment]
```
vec_struct: MACRO
.X:: ds 1
.Y:: ds 1
.Z:: ds 1
ENDM

wPosition:: vec_struct
```Does pretty much the same thing.

**Comment by Rangi42** [comment]
So I'm realizing that `__SCOPE__` would be the cleanest solution for these macros to verify data table sizes. I'd like to write this:

```
ThingTable:
    table_width 2
    dw Foo
    dw Bar
    ...
    dw Last
    check_table_length NUM_THINGS
```

ideally without typing "`ThingTable`" twice, without the macros creating any labels of their own, and without requiring the section's address to be fixed. (`check_table_length` uses an `assert` which can be evaluated at link time.)

This is not a use case that rgbasm-native structs will eventually solve. Without `__SCOPE__` I would either have to:

1. write `table_width 2, ThingTable` and store the starting label as an `EQUS`. This avoids creating extra labels, and gives me a nice value to be put in an assertion failure message, but means repeating the name twice.
2. make `table_width` generate a new starting label (`.TableStart\@`) and still store that in an `EQUS` (so `check_table_length` can refer to it). This avoids repetition, but has limits, whether the generated label is local or global.

With `__SCOPE__` I could just do:

```
table_width: MACRO
REDEF __CURRENT_TABLE_START EQUS "\1"
__CURRENT_TABLE_WIDTH = \2
ENDM

check_table_length: MACRO
n = \1
    assert n * __CURRENT_TABLE_WIDTH == @ - __CURRENT_TABLE_START, \
        "Expected {d:n} entries, each {d:__CURRENT_TABLE_WIDTH} bytes"
ENDM
```

**Comment by ISSOtm** [comment]
I understand restrictions 1 and 3, but 2 is arbitrary. I posted macros that define the labels in a greppable way with no redundancy.

**Comment by Rangi42** [comment]
If you mean passing the label *definition* as a macro parameter, like `table ThingTable:, 2`, it technically works but colon-comma looks unnatural. To me `__SCOPE__` seems like exposing the same kind of pre-existing internal state as `__LINE__`. (Neither of them is needed often, and both get printed in ordinary error traces anyway, but that doesn't mean `__LINE__` will be deprecated.)

**Comment by daid** [comment]
Still no need for `__SCOPE__`

https://daid.github.io/rgbds-live/#LYQwlgdgdCDOwAAAuIBGAbApgfQO5gBMkALALgAIBZAQQGEAlAeQFgAoKbbFDHWFAJySk2nWgFV69AKIA5ACrY51AEIAZKdgDqASQAicgBLkAvOQA6ARjazdlNmwDGxTA4DWXNFmxYIAcxIUNAwsrBAm5las5NHkcLCYguRhAFTkohLS8ooq6lp6hiamAALkALTkHB482HwgggA05mwxLeQARFIAHgAOLkiYBOQA3gSkEAC+5JgQSPxgmLCNmCBOw6PpkrIKSmoaOvoGk6gAnv2wbdYytvasAMpStHLajDLtZ0htjUyUAAxscsRIL45J5MMIojFuF58ERiOQAEzNGIEXDkABiAHsMUjoijyMo6jjyHjVHAkESnC53FCcD5-HCACxEtiYjHggn8cGkvikBBAA

**Comment by Rangi42** [comment]
> Still no need for `__SCOPE__`

Yes, I'm doing something similar for now:

```
table_width: MACRO
CURRENT_TABLE_WIDTH = \1
if DEF(CURRENT_TABLE_START)
PURGE CURRENT_TABLE_START
endc
if _NARG == 2
CURRENT_TABLE_START EQUS "\2"
else
CURRENT_TABLE_START EQUS "._table_width\@"
CURRENT_TABLE_START:
endc
ENDM

assert_table_length: MACRO
x = \1
	assert x * CURRENT_TABLE_WIDTH == @ - CURRENT_TABLE_START, \
		"{CURRENT_TABLE_START}: expected {d:x} entries, each {d:CURRENT_TABLE_WIDTH} bytes"
ENDM
```

But clearly it doesn't mean there's "no need" for `__SCOPE__`, since it solves none of the problems I mentioned with various workarounds (this either creates an arbitrary local label or has to be redundantly passed an existing label, and in the former case the assertion string can't mention the table's label name).

**Comment by Rangi42** [comment]
What if `__SCOPE__` just expanded to an empty string when there's no global label? This would solve @ISSOtm 's initial critique about going to extra trouble to make `__SCOPE__` be un`DEF`ined in that case. It would also make it more useful in the same contexts as `__FILE__` and `__LINE__`: debugging errors could then start with `STRCAT(__FILE__, ":{d:__LINE__}:{__SCOPE__}:", "some error message")`. And instead of checking `if DEF(__SCOPE__)`, check `if !STRLEN("{__SCOPE__}")`.

**Comment by aaaaaa123456789** [comment]
"There's no need for feature X because you can do <simple workaround that only has half the features> or <very complicated workaround that maybe works, but is really not practical>" is a statement against `__SCOPE__` as much as it is against half the features in the codebase.

You don't _need_ `warn` and `fail`; you can print messages and (in the latter case) force building to fail. You don't _need_ `rb` and friends; you can emulate them with macros. You don't _need_ `elif`; you can use an `if` inside an `else`. And so on, and so forth.

**Comment by AntonioND** [comment]
> "There's no need for feature X because you can do <simple workaround that only has half the features> or <very complicated workaround that maybe works, but is really not practical>" is a statement against `__SCOPE__` as much as it is against half the features in the codebase.

Exactly.

**Comment by Rangi42** [comment]
For what it's worth, two Python scripts I've written to analyze asm code both keep track of the current label scope and print it out in messages along with the file+line. So I expect `__SCOPE__` would be a useful addition to whatever debug macros people are writing that previously used `__FILE__` and `__LINE__`.

**Comment by Rangi42** [comment]
@ISSOtm Is the update without a hardcoded check in `sym_FindExactSymbol` acceptable?

**Comment by ISSOtm** [comment]
I agree that many of RGBASM's features are not strictly necessary, but are convenient and improve the code quality significantly.

The key part is "significantly": balancing the benefits gained with the increased complexity (both in maintaining and, possibly, performance); and it's largely subjective.
I believe that a non-negligible amount of RGBASM features don't match that balance, sometimes because of legacy, sometimes because of decisions that I came to regret later. But I don't think rolling back would be worth it, so I'm keeping them.
(Yes, I think RGBASM is slightly bloated. We all make mistakes, so imo it's acceptable.)

You're welcome to disagree, but I don't think this change is useful enough to warrant native syntax/support.

Further, it severely limits future expansions to scoping (which would be very useful), and I'm afraid about its interactions with other features. (Think #342, for example.)

**tl;dr:** I don't like this feature as-is, so I'm going to put it on the backburner for now, and will look at it after more important features have been dealt with. (See what's attached to the [0.5.1 milestone](https://github.com/gbdev/rgbds/milestone/6), basically.)

**Comment by Rangi42** [comment]
As long as `.local` labels exist, "the currently scoped label" should also exist; I don't expect the future plans for more complex scoping (like auto-purged symbols local to a file/macro/rept, or local names that shadow outer ones for macros' or user-defined functions' parameters) to affect that. I agree `__SCOPE__` would have limited applications, just not that it would limit future changes.

**Comment by ISSOtm** [comment]
I'm considering more complex scoping like ca65's `Outer.inner.itsDarkInHere`.

**Comment by Rangi42** [comment]
(For `Outer.inner.itsDarkInHere` I'd expect `__SCOPE__` to be `Outer.inner`, but regardless, that's beyond this particular implementation. I'll see if something like this is useful+practical depending on future scope-related changes.)
